sympy/core/basic.py

```diff
@@ -1963,6 +1963,10 @@ def _sage_(self):
     def could_extract_minus_sign(self):
         return False  # see Expr.could_extract_minus_sign
 
+    @classmethod
+    def _evaluate(cls, expr):
+        return expr
```

Comment:
I don't think that this method should be added to Basic. There are already too many methods on Basic that are added haphazardly and then used by only a tiny handful of subclasses. New methods should not be added without defining and documenting what the overall design is supposed to be.

Reply:
This is related to the well-known problem of SymPy that it is not well defined when expressions evaluate and when they do not. `.expand( )` on matrix expressions does not completely expand, as the constructors in the matrix expressions module do not evaluate by default. This behaviour makes sense (I would rather change the behaviour of the rest of SymPy, the matrix expressions module has a robust design).

Maybe singledispatch could avoid the problem of too many overloaded methods in `Basic` ?

Reply:
I don't think singledispatch vs methods really makes any difference. The point is really that every time core logic gets complicated by needing to call additional functions/methods it becomes hard to improve that core logic in future e.g. to make it faster or simpler or more consistent.

Here if the goal is to change the behaviour of `MatAdd.expand` then why isn't the logic in the `MatAdd.expand` function? At least that would localise it to where it is needed.

Reply:
> Here if the goal is to change the behaviour of `MatAdd.expand` then why isn't the logic in the `MatAdd.expand` function? At least that would localise it to where it is needed.

That could be an option. Though, are we giving up completely the idea of not evaluating all `__new__` constructors by default?